src/libsyntax/parse/parser.rs

```diff
@@ -1744,30 +1744,42 @@ impl<'a> Parser<'a> {
     fn parse_arg_general(&mut self, require_name: bool) -> PResult<'a, Arg> {
         maybe_whole!(self, NtArg, |x| x);
 
-        let (pat, ty) = if require_name || self.is_named_argument() {
-            debug!("parse_arg_general parse_pat (require_name:{})",
-                   require_name);
-            let pat = self.parse_pat()?;
+        let parser_snapshot_before_pat = self.clone();
 
+        // We're going to try parsing the argument as a pattern (even if it's not
+        // allowed, such as for trait methods without bodies). This way we can provide
+        // better errors to the user.
+        let pat_arg: PResult<'a, (P<Pat>, P<Ty>)> = do catch {
+            let pat = self.parse_pat()?;
             self.expect(&token::Colon)?;
             (pat, self.parse_ty()?)
-        } else {
-            debug!("parse_arg_general ident_to_pat");
-            let ident = Ident::new(keywords::Invalid.name(), self.prev_span);
-            let ty = self.parse_ty()?;
-            let pat = P(Pat {
-                id: ast::DUMMY_NODE_ID,
-                node: PatKind::Ident(BindingMode::ByValue(Mutability::Immutable), ident, None),
-                span: ty.span,
-            });
-            (pat, ty)
         };
 
-        Ok(Arg {
-            ty,
-            pat,
-            id: ast::DUMMY_NODE_ID,
-        })
+        let is_named_argument = self.is_named_argument();
+        match pat_arg {
+            Ok((pat, ty)) => {
+                Ok(Arg { ty, pat, id: ast::DUMMY_NODE_ID })
+            }
+            Err(mut err) => {
+                if require_name || is_named_argument {
+                    Err(err)
+                } else {
+                    err.cancel();
+                    // Recover from attempting to parse the argument as a pattern. This means
+                    // the type is alone, with no name, e.g. `fn foo(u32)`.
+                    mem::replace(self, parser_snapshot_before_pat);
```

Comment:
Can this be checked without rewinding state?proactively looking if a path could be parsed? I'm not against the rewind method in exceptional cases, but it should be a tool of last resort.

Reply:
I had also wanted to avoid this, but I wasn't quite sure how to go about it. The problem I had was that `parse_pat` and `parse_ty` can succeed on the same strings. Ideally we'd `parse_pat` (which should be a superset of `parse_ty`), try to parse a `:` and then convert the pattern to a type if it failed, because we effectively have all the information we need at that stage. Is there any way to reinterpret a pattern as a type, or similar? (Or perhaps there's a simpler method I'm overlooking.)

Reply:
That seems plausible. I'll give it a go soon.

Reply:
Rewinding is rare, so it's not very useful to optimize for that, but doing the state *saving* for every single fn parameter in traits doesn't look good.
I think doing minimal lookahead and avoiding state saving for the common case `ident: Type` would be enough.